tray/monitor.go

```diff
@@ -0,0 +1,294 @@
+package tray
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/NordSecurity/nordvpn-linux/cli"
+	"github.com/NordSecurity/nordvpn-linux/daemon/pb"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	meshpb "github.com/NordSecurity/nordvpn-linux/meshnet/pb"
+
+	"github.com/NordSecurity/systray"
+	"github.com/fatih/color"
+	"google.golang.org/grpc/status"
+)
+
+type stateType struct {
+	daemonAvailable bool
+	loggedIn        bool
+	vpnActive       bool
+	meshnetEnabled  bool
+	daemonError     string
+	accountName     string
+	vpnStatus       string
+	vpnHostname     string
+	vpnCity         string
+	vpnCountry      string
+	mu              sync.RWMutex
+}
+
+var state = stateType{vpnStatus: "Disconnected"}
+
+// The pattern is to return 'true' if something has changed and 'false' when no changes were detected
+
+func ping(client pb.DaemonClient) bool {
+	changed := false
+	daemonAvailable := false
+	daemonError := ""
+
+	resp, err := client.Ping(context.Background(), &pb.Empty{})
+	if err != nil {
+		daemonError = internal.ErrDaemonConnectionRefused.Error()
+		if strings.Contains(err.Error(), "no such file or directory") {
+			daemonError = "nordvpnd is not running"
+		}
+		if strings.Contains(err.Error(), "permission denied") {
+			daemonError = "add a user to the nordvpn group"
+		}
+		if snapErr := cli.RetrieveSnapConnsError(err); snapErr != nil {
+			daemonError = cli.FormatSnapMissingConnsErr(snapErr)
+		}
+	} else {
+		switch resp.Type {
+		case internal.CodeOffline:
+			daemonError = cli.ErrInternetConnection.Error()
+		case internal.CodeDaemonOffline:
+			daemonError = internal.ErrDaemonConnectionRefused.Error()
+		case internal.CodeOutdated:
+			daemonError = cli.ErrUpdateAvailable.Error()
+		}
+	}
+
+	if daemonError == "" {
+		daemonAvailable = true
+	}
+
+	state.mu.Lock()
+
+	if !state.daemonAvailable && daemonAvailable {
+		state.daemonAvailable = true
+		changed = true
+		if NotifyEnabled {
+			defer notification("info", "Connected to NordVPN daemon")
+		}
+	} else if state.daemonAvailable && !daemonAvailable {
+		state.daemonAvailable = false
+		changed = true
+		if NotifyEnabled {
+			defer notification("info", "Disconnected from NordVPN daemon")
+		}
+	}
+
+	if state.daemonError != daemonError {
+		state.daemonError = daemonError
+		changed = true
+	}
+
+	state.mu.Unlock()
+	return changed
+}
+
+func fetchLogged(client pb.DaemonClient) bool {
```

Comment:
Confusing function. Name implies that it fetches logged in state and should return whether user is logged in but does multiple things:
1. Fetches state from daemon
2. Updates the local state
3. Sends user facing notification
4. Informs whether state changed with return value
Could this be refactored in any way?